File: apps/web/lib/api/services/contextual-compare.ts
/**
 * Contextual Comparison API Service
 * 
 * "Context vs Context" comparison - the differentiator.
 */

import { apiClient } from '../client';

// =============================================================================
// TYPES
// =============================================================================

export interface SimilarityBreakdown {
  duration: number;
  intensity: number;
  avg_hr: number;
  max_hr: number;
  type: number;
  conditions: number;
  elevation: number;
}

export interface SplitData {
  split_number: number;
  distance_m: number;
  elapsed_time_s: number;
  pace_per_km: number | null;
  avg_hr: number | null;
  max_hr: number | null;
  cadence: number | null;
  gap_per_mile: number | null;  // Grade Adjusted Pace
  efficiency: number | null;     // speed / HR for this split
  cumulative_distance_m: number;
}

export interface AdvancedAnalytics {
  cardiac_drift_pct: number | null;      // HR increase over run (%)
  aerobic_decoupling_pct: number | null; // Efficiency drop first vs second half (%)
  pace_variability_cv: number | null;    // Coefficient of variation (%)
  fade_pct: number | null;               // Second half slower than first (%)
  first_half_pace: number | null;        // sec/km
  second_half_pace: number | null;       // sec/km
  first_half_hr: number | null;          // bpm
  second_half_hr: number | null;         // bpm
  first_half_efficiency: number | null;
  second_half_efficiency: number | null;
  split_type: 'negative' | 'positive' | 'even' | null;
}

export interface SimilarRun {
  id: string;
  date: string;
  name: string;
  workout_type: string | null;
  distance_m: number;
  distance_km: number;
  duration_s: number;
  pace_per_km: number | null;
  pace_formatted: string | null;
  avg_hr: number | null;
  max_hr: number | null;
  efficiency: number | null;
  intensity_score: number | null;
  elevation_gain: number | null;
  temperature_f: number | null;
  similarity_score: number;
  similarity_breakdown: SimilarityBreakdown;
  splits: SplitData[];
}

export interface GhostAverage {
  num_runs_averaged: number;
  avg_pace_per_km: number | null;
  avg_pace_formatted: string | null;
  avg_hr: number | null;
  avg_max_hr: number | null;
  avg_efficiency: number | null;
  avg_duration_s: number | null;
  avg_duration_formatted: string | null;
  avg_distance_m: number | null;
  avg_distance_km: number | null;
  avg_elevation_gain: number | null;
  avg_temperature_f: number | null;
  avg_intensity_score: number | null;
  avg_splits: {
    split_number: number;
    avg_pace_per_km: number | null;
    avg_hr: number | null;
    num_runs_at_split: number;
  }[];
}

export interface PerformanceScore {
  score: number;
  rating: 'exceptional' | 'strong' | 'solid' | 'average' | 'below' | 'struggling';
  percentile: number;
  pace_vs_baseline: number;
  efficiency_vs_baseline: number;
  hr_vs_baseline: number;
  age_graded_performance: number | null;
}

export interface ContextFactor {
  name: string;
  icon: string;
  this_run_value: string;
  baseline_value: string;
  difference: string;
  impact: 'positive' | 'negative' | 'neutral';
  explanation: string;
}

export interface TargetRun {
  id: string;
  date: string;
  name: string;
  workout_type: string | null;
  workout_zone: string | null;
  distance_m: number;
  distance_km: number;
  duration_s: number;
  duration_formatted: string | null;
  pace_per_km: number | null;
  pace_formatted: string | null;
  avg_hr: number | null;
  max_hr: number | null;
  efficiency: number | null;
  intensity_score: number | null;
  elevation_gain: number | null;
  average_speed_kmh: number | null;
  temperature_f: number | null;
  humidity_pct: number | null;
  weather_condition: string | null;
  performance_percentage: number | null;
  performance_percentage_national: number | null;
  splits: SplitData[];
  analytics: AdvancedAnalytics;
}

export interface ContextualComparisonResult {
  target_run: TargetRun;
  similar_runs: SimilarRun[];
  ghost_average: GhostAverage;
  performance_score: PerformanceScore;
  context_factors: ContextFactor[];
  headline: string;
  key_insight: string;
}

export interface QuickScore {
  activity_id: string;
  score: number | null;
  rating: string | null;
  headline: string;
  similar_runs_count: number;
  key_insight: string | null;
}

// =============================================================================
// API FUNCTIONS
// =============================================================================

/**
 * Find similar runs and build contextual comparison
 */
export async function findSimilarRuns(
  activityId: string,
  options: {
    maxResults?: number;
    minSimilarity?: number;
    daysBack?: number;
  } = {}
): Promise<ContextualComparisonResult> {
  const params = new URLSearchParams();
  if (options.maxResults) params.set('max_results', options.maxResults.toString());
  if (options.minSimilarity) params.set('min_similarity', options.minSimilarity.toString());
  if (options.daysBack) params.set('days_back', options.daysBack.toString());
  
  const queryString = params.toString();
  const url = `/v1/compare/similar/${activityId}${queryString ? `?${queryString}` : ''}`;
  
  return apiClient.get<ContextualComparisonResult>(url);
}

/**
 * Auto-compare with most similar runs (one-click magic)
 */
export async function autoCompareSimilar(
  activityId: string
): Promise<ContextualComparisonResult> {
  return apiClient.get<ContextualComparisonResult>(`/v1/compare/auto-similar/${activityId}`);
}

/**
 * Compare user-selected runs
 */
export async function compareSelectedRuns(
  activityIds: string[],
  baselineId?: string
): Promise<ContextualComparisonResult> {
  return apiClient.post<ContextualComparisonResult>('/v1/compare/selected', {
    activity_ids: activityIds,
    baseline_id: baselineId,
  });
}

/**
 * Get quick performance score for an activity
 */
export async function getQuickScore(
  activityId: string
): Promise<QuickScore> {
  return apiClient.get<QuickScore>(`/v1/compare/quick-score/${activityId}`);
}

// =============================================================================
// HR-BASED SEARCH FUNCTIONS
// Explicit heart rate based queries - foundation for query engine
// =============================================================================

export interface HRSearchActivity {
  id: string;
  name: string;
  date: string;
  workout_type: string | null;
  distance_m: number | null;
  distance_km: number | null;
  duration_s: number | null;
  pace_per_km: number | null;
  pace_formatted: string | null;
  avg_hr: number | null;
  max_hr: number | null;
  intensity_score: number | null;
  temperature_f: number | null;
  elevation_gain: number | null;
  hr_diff?: number; // Difference from target avg HR
  max_hr_diff?: number; // Difference from target max HR
}

export interface HRSearchResult {
  target_activity_id?: string;
  hr_tolerance?: number;
  hr_range?: { min: number; max: number };
  min_duration_minutes?: number | null;
  count: number;
  activities: HRSearchActivity[];
}

/**
 * Find runs with similar average heart rate
 */
export async function findByAvgHR(
  activityId: string,
  options: {
    tolerance?: number;
    maxResults?: number;
    daysBack?: number;
  } = {}
): Promise<HRSearchResult> {
  const params = new URLSearchParams();
  if (options.tolerance) params.set('tolerance', options.tolerance.toString());
  if (options.maxResults) params.set('max_results', options.maxResults.toString());
  if (options.daysBack) params.set('days_back', options.daysBack.toString());
  
  const queryString = params.toString();
  const url = `/v1/compare/by-avg-hr/${activityId}${queryString ? `?${queryString}` : ''}`;
  
  return apiClient.get<HRSearchResult>(url);
}

/**
 * Find runs with similar maximum heart rate
 */
export async function findByMaxHR(
  activityId: string,
  options: {
    tolerance?: number;
    maxResults?: number;
    daysBack?: number;
  } = {}
): Promise<HRSearchResult> {
  const params = new URLSearchParams();
  if (options.tolerance) params.set('tolerance', options.tolerance.toString());
  if (options.maxResults) params.set('max_results', options.maxResults.toString());
  if (options.daysBack) params.set('days_back', options.daysBack.toString());
  
  const queryString = params.toString();
  const url = `/v1/compare/by-max-hr/${activityId}${queryString ? `?${queryString}` : ''}`;
  
  return apiClient.get<HRSearchResult>(url);
}

/**
 * Find all runs within an average HR range (no reference activity needed)
 */
export async function findByHRRange(
  minHR: number,
  maxHR: number,
  options: {
    minDuration?: number; // in minutes
    maxResults?: number;
    daysBack?: number;
  } = {}
): Promise<HRSearchResult> {
  const params = new URLSearchParams();
  params.set('min_hr', minHR.toString());
  params.set('max_hr', maxHR.toString());
  if (options.minDuration) params.set('min_duration', options.minDuration.toString());
  if (options.maxResults) params.set('max_results', options.maxResults.toString());
  if (options.daysBack) params.set('days_back', options.daysBack.toString());
  
  return apiClient.get<HRSearchResult>(`/v1/compare/hr-range?${params.toString()}`);
}

// =============================================================================
// METRIC HISTORY (for interactive tiles)
// =============================================================================

export interface MetricHistoryEntry {
  activity_id: string;
  date: string;
  name: string;
  efficiency: number | null;
  cardiac_drift: number | null;
  aerobic_decoupling: number | null;
  pace_consistency: number | null;
}

export interface MetricStats {
  avg: number | null;
  best: number | null;
  worst: number | null;
}

export interface MetricHistoryResult {
  activity_id: string;
  history_count: number;
  current: {
    efficiency: number | null;
    cardiac_drift: number | null;
    aerobic_decoupling: number | null;
    pace_consistency: number | null;
  };
  statistics: {
    efficiency: MetricStats;
    cardiac_drift: MetricStats;
    aerobic_decoupling: MetricStats;
    pace_consistency: MetricStats;
  };
  history: MetricHistoryEntry[];
  insights: {
    efficiency?: string;
    cardiac_drift?: string;
    aerobic_decoupling?: string;
    pace_consistency?: string;
  };
}

/**
 * Get historical metric data for interactive tile drill-down
 */
export async function getMetricHistory(activityId: string): Promise<MetricHistoryResult> {
  return apiClient.get<MetricHistoryResult>(`/v1/compare/metric-history/${activityId}`);
}

File: apps/api/routers/contextual_compare.py
"""
Contextual Comparison API

The differentiator: "Context vs Context" comparison.

Endpoints:
- GET /v1/compare/similar/{activity_id} - Find similar runs and compare
- POST /v1/compare/selected - Compare user-selected runs
- GET /v1/compare/auto-similar/{activity_id} - Auto-find most similar runs (no selection)
"""

from typing import List, Optional
from uuid import UUID
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from core.database import get_db
from core.auth import get_current_user
from models import Athlete
from services.contextual_comparison import ContextualComparisonService

router = APIRouter(prefix="/v1/compare", tags=["compare"])


# =============================================================================
# REQUEST/RESPONSE MODELS
# =============================================================================

class SelectedCompareRequest(BaseModel):
    """Request to compare user-selected activities"""
    activity_ids: List[str]  # UUIDs as strings
    baseline_id: Optional[str] = None  # Which activity to use as the baseline


class SimilarityConfig(BaseModel):
    """Configuration for similarity matching"""
    max_results: int = 10
    min_similarity: float = 0.3
    days_back: int = 365


# =============================================================================
# ENDPOINTS
# =============================================================================

@router.get("/similar/{activity_id}")
def find_similar_runs(
    activity_id: UUID,
    max_results: int = Query(10, ge=1, le=20),
    min_similarity: float = Query(0.3, ge=0.0, le=1.0),
    days_back: int = Query(365, ge=30, le=730),
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Find runs similar to the target activity and build a contextual comparison.
    
    This is the core "Context vs Context" feature.
    
    Returns:
    - target_run: The activity being analyzed
    - similar_runs: List of similar runs with similarity scores
    - ghost_average: The baseline "ghost" averaged from similar runs
    - performance_score: How the target compares to the ghost
    - context_factors: Explanations for performance differences
    - headline: Plain-language performance summary
    - key_insight: The "BUT" explanation
    """
    service = ContextualComparisonService(db)
    
    try:
        result = service.find_similar_runs(
            activity_id=activity_id,
            athlete_id=current_user.id,
            max_results=max_results,
            min_similarity=min_similarity,
            days_back=days_back,
        )
        return result.to_dict()
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/selected")
def compare_selected_runs(
    request: SelectedCompareRequest,
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Compare user-selected runs with contextual analysis.
    
    Use this when the athlete manually selects 2-10 runs to compare.
    
    If baseline_id is provided, that run is treated as the "target" being analyzed.
    Otherwise, the most recent run is the target.
    """
    # Convert string IDs to UUIDs
    try:
        activity_ids = [UUID(id_str) for id_str in request.activity_ids]
        baseline_id = UUID(request.baseline_id) if request.baseline_id else None
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid activity ID format")
    
    service = ContextualComparisonService(db)
    
    try:
        result = service.compare_selected_runs(
            activity_ids=activity_ids,
            athlete_id=current_user.id,
            baseline_id=baseline_id,
        )
        return result.to_dict()
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/auto-similar/{activity_id}")
def auto_compare_similar(
    activity_id: UUID,
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    One-click contextual comparison.
    
    Automatically finds the 10 most similar runs and builds a complete
    comparison with ghost average, performance score, and insights.
    
    This is the "magic" endpoint - minimal input, maximum insight.
    """
    service = ContextualComparisonService(db)
    
    try:
        result = service.find_similar_runs(
            activity_id=activity_id,
            athlete_id=current_user.id,
            max_results=10,
            min_similarity=0.25,  # Slightly lower threshold for auto
            days_back=365,
        )
        return result.to_dict()
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/quick-score/{activity_id}")
def get_quick_performance_score(
    activity_id: UUID,
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Get just the performance score for an activity.
    
    Lighter-weight endpoint for showing scores on activity lists.
    Returns: score, rating, headline
    """
    service = ContextualComparisonService(db)
    
    try:
        result = service.find_similar_runs(
            activity_id=activity_id,
            athlete_id=current_user.id,
            max_results=5,  # Fewer for speed
            min_similarity=0.3,
            days_back=180,  # Shorter window for speed
        )
        
        return {
            "activity_id": str(activity_id),
            "score": result.performance_score.score,
            "rating": result.performance_score.rating,
            "headline": result.headline,
            "similar_runs_count": len(result.similar_runs),
            "key_insight": result.key_insight,
        }
    except ValueError:
        return {
            "activity_id": str(activity_id),
            "score": None,
            "rating": None,
            "headline": "Not enough similar runs for comparison",
            "similar_runs_count": 0,
            "key_insight": None,
        }


# =============================================================================
# EXPLICIT HR SEARCH ENDPOINTS
# These allow athletes to explicitly search/filter by heart rate.
# Foundation for the upcoming query engine.
# =============================================================================

@router.get("/by-avg-hr/{activity_id}")
def find_by_avg_hr(
    activity_id: UUID,
    tolerance: int = Query(5, ge=1, le=20, description="HR tolerance in bpm (±)"),
    max_results: int = Query(20, ge=1, le=50),
    days_back: int = Query(365, ge=30, le=730),
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Find runs with similar average heart rate.
    
    Physiologically, same avg HR = same relative effort level.
    Use this to find runs where you maintained similar effort regardless of pace.
    
    Example: "Show me all runs where I held ~145 bpm"
    """
    service = ContextualComparisonService(db)
    
    results = service.find_by_avg_hr(
        activity_id=activity_id,
        athlete_id=current_user.id,
        hr_tolerance=tolerance,
        max_results=max_results,
        days_back=days_back,
    )
    
    return {
        "target_activity_id": str(activity_id),
        "hr_tolerance": tolerance,
        "count": len(results),
        "activities": results,
    }


@router.get("/by-max-hr/{activity_id}")
def find_by_max_hr(
    activity_id: UUID,
    tolerance: int = Query(5, ge=1, le=20, description="HR tolerance in bpm (±)"),
    max_results: int = Query(20, ge=1, le=50),
    days_back: int = Query(365, ge=30, le=730),
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Find runs with similar maximum heart rate.
    
    Same max HR indicates similar cardiovascular peak stress.
    Use this to find runs where you hit similar intensity peaks.
    
    Example: "Show me runs where I peaked at ~175 bpm"
    """
    service = ContextualComparisonService(db)
    
    results = service.find_by_max_hr(
        activity_id=activity_id,
        athlete_id=current_user.id,
        hr_tolerance=tolerance,
        max_results=max_results,
        days_back=days_back,
    )
    
    return {
        "target_activity_id": str(activity_id),
        "hr_tolerance": tolerance,
        "count": len(results),
        "activities": results,
    }


@router.get("/hr-range")
def find_by_hr_range(
    min_hr: int = Query(..., ge=60, le=220, description="Minimum average HR"),
    max_hr: int = Query(..., ge=60, le=220, description="Maximum average HR"),
    min_duration: Optional[int] = Query(None, ge=1, description="Minimum duration in minutes"),
    max_results: int = Query(50, ge=1, le=100),
    days_back: int = Query(365, ge=30, le=730),
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Find all runs within an average HR range.
    
    Pure query endpoint - no reference activity needed.
    Foundation primitive for the query engine.
    
    Example: "Show me all runs where I maintained 145-155 bpm for 30+ minutes"
    """
    if min_hr > max_hr:
        raise HTTPException(status_code=400, detail="min_hr must be <= max_hr")
    
    service = ContextualComparisonService(db)
    
    results = service.find_by_hr_range(
        athlete_id=current_user.id,
        min_hr=min_hr,
        max_hr=max_hr,
        min_duration_minutes=min_duration,
        max_results=max_results,
        days_back=days_back,
    )
    
    return {
        "hr_range": {"min": min_hr, "max": max_hr},
        "min_duration_minutes": min_duration,
        "count": len(results),
        "activities": results,
    }


@router.get("/metric-history/{activity_id}")
def get_metric_history(
    activity_id: UUID,
    current_user: Athlete = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """
    Get historical metric data for the interactive tile drill-down.
    
    Returns efficiency, cardiac drift, aerobic decoupling, and pace consistency
    trends from similar runs to provide context for the current run's metrics.
    """
    service = ContextualComparisonService(db)
    
    try:
        history = service.get_metric_history(
            activity_id=activity_id,
            athlete_id=current_user.id,
        )
        return history
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
